### PR TITLE
Implement ContinuousBCQPolicy and offline_bcq example

### PR DESCRIPTION
**Pull request by thkkk**
- [x] I have marked all applicable categories:
    + [ ] exception-raising fix
    + [ ] algorithm implementation fix
    + [ ] documentation modification
    + [x] new feature
- [x] I have reformatted the code using `make format` (**required**)
- [x] I have checked the code using `make commit-checks` (**required**)
- [ ] If applicable, I have mentioned the relevant/related issue(s)
- [x] If applicable, I have listed every items in this Pull Request below

This PR implements ContinuousBCQPolicy, which could be used to train offline agent in the environment of continuous action space. Here is an experimental result in 'halfcheetah-expert-v1', a d4rl environment (for Offline Reinforcement Learning).

image

Example usage is in the `examples/offline/offline_bcq.py`.

### CODE REVIEW

**Comment by Trinkle23897** [comment]
Could you please add `test/continuous/bcq.py` for BCQ unit test?

**Review by thkkk — COMMENTED** [review]
finish test_bcq

File: examples/offline/README.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+
+
+## Results
+
+| Environment \\ Algorithm | BCQ           |      |
+| ------------------------ | ------------- | ---- |
+| halfcheetah-expert-v1    | 10624.0±181.4 |      |
+|                          |               |      |
+
```
Comment (Trinkle23897) [inline examples/offline/README.md]:
```suggestion
## Results

| Environment \\ Algorithm | BCQ             |
| ------------------------ | --------------- |
| halfcheetah-expert-v1    | 10624.0 ± 181.4 |

```

File: examples/offline/offline_bcq.py [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+import d4rl
+import gym
+import numpy as np
+import torch
+from torch.utils.tensorboard import SummaryWriter
+
+from tianshou.data import Batch, Collector, ReplayBuffer, VectorReplayBuffer
+from tianshou.env import SubprocVectorEnv
+from tianshou.policy import BCQPolicy
+from tianshou.policy.imitation.bcq import VAE, Perturbation
```
Comment (Trinkle23897) [inline examples/offline/offline_bcq.py]: Do you think moving `VAE` and `Pertubation` to `utils/net/...` is a good idea or not? Because previously we all place our network architecture under that module.

Comment (Trinkle23897) [reply examples/offline/offline_bcq.py]: BTW it's easier for other algorithms to reuse VAE, if needed.

File: examples/offline/offline_bcq.py [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+
+    # vae
+    # args.vae_hidden_sizes = [750, 750]
+    # output_dim = 0, so the last Module in the encoder is ReLU
+    vae_encoder = MLP(
+        input_dim=args.state_dim + args.action_dim,
+        hidden_sizes=args.vae_hidden_sizes,
+        device=args.device
+    )
+    args.latent_dim = args.action_dim * 2
```
Comment (Trinkle23897) [inline examples/offline/offline_bcq.py]: Could you please explain why *2 here?

File: examples/offline/offline_bcq.py [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+        torch.save(policy.state_dict(), os.path.join(log_path, 'policy.pth'))
+
+    def watch():
+        policy_path = args.resume_path \
+            if args.resume_path is not None \
+            else os.path.join(log_path, 'policy.pth')
+
+        policy.load_state_dict(
+            torch.load(policy_path, map_location=torch.device('cpu'))
+        )  # log_path,
```
Comment (Trinkle23897) [inline examples/offline/offline_bcq.py]: Can we do this?
```suggestion
        if args.resume_path is None:
            args.resume_path = os.path.join(log_path, 'policy.pth')

        policy.load_state_dict(
            torch.load(args.resume_path, map_location=torch.device('cpu'))
        )  # log_path,
```

File: test/continuous/test_bcq.py [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+    parser.add_argument('--alpha', type=float, default=0.2)
+    parser.add_argument('--auto_alpha', type=int, default=1)
+    parser.add_argument('--alpha_lr', type=float, default=3e-4)
+    parser.add_argument('--rew_norm', action="store_true", default=False)
+    parser.add_argument('--n_step', type=int, default=3)
+    args = parser.parse_known_args()[0]
+    return args
+
+
+def gather_data():
```
Comment (Trinkle23897) [inline test/continuous/test_bcq.py]: This part has a lot of duplicated code. I suggest reusing the existing code (and also using some features of pytest) to finish data generation (save to .pkl / .h5 format and load from that dataset).

File: tianshou/policy/__init__.py [hunk 6]
```diff
@@ -45,6 +46,7 @@
     "DiscreteSACPolicy",
     "ImitationPolicy",
     "DiscreteBCQPolicy",
+    "BCQPolicy",
```
Comment (Trinkle23897) [inline tianshou/policy/__init__.py]:
```suggestion
    "BCQPolicy",
    "DiscreteBCQPolicy",
```

File: tianshou/policy/imitation/bcq.py [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+    :param torch.nn.Module decoder: the decoder in vae. Its input_dim must be
+        state_dim + action_dim, and output_dim must be action_dim.
+    :param int hidden_dim: the size of the last linear-layer in encoder.
+    :param int latent_dim: the size of latent layer.
+    :param float max_action: the maximum value of each dimension of action.
+    :param Union[str, torch.device] device: which device to create this model on.
+        Default to cpu.
+
+    .. seealso::
+        You can refer to `examples/offline/offline_bcq.py` to see how to use it.
```
Comment (Trinkle23897) [inline tianshou/policy/imitation/bcq.py]:
```suggestion
    .. seealso::

        You can refer to `examples/offline/offline_bcq.py` to see how to use it.
```

File: tianshou/policy/imitation/bcq.py [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+
+    :param torch.nn.Module encoder: the encoder in vae. Its input_dim must be
+        state_dim + action_dim, and output_dim must be hidden_dim.
+    :param torch.nn.Module decoder: the decoder in vae. Its input_dim must be
+        state_dim + action_dim, and output_dim must be action_dim.
+    :param int hidden_dim: the size of the last linear-layer in encoder.
+    :param int latent_dim: the size of latent layer.
+    :param float max_action: the maximum value of each dimension of action.
+    :param Union[str, torch.device] device: which device to create this model on.
+        Default to cpu.
```
Comment (Trinkle23897) [inline tianshou/policy/imitation/bcq.py]:
```suggestion
        Default to "cpu".
```

File: tianshou/policy/imitation/bcq.py [hunk 9]
```diff
@@ -0,0 +1,326 @@
...
+    def forward(self, state: torch.Tensor, action: torch.Tensor) -> torch.Tensor:
+        # preprocess_net
+        logits = self.preprocess_net(torch.cat([state, action], 1))[0]
+        a = self.phi * self.max_action * torch.tanh(logits)
+        # clip to [-max_action, max_action]
+        return (a + action).clamp(-self.max_action, self.max_action)
+
+
+class VAE(nn.Module):
+    """Implementation of vae in BCQ algorithm.
```
Comment (Trinkle23897) [inline tianshou/policy/imitation/bcq.py]:
```suggestion
    """Implementation of VAE in BCQ algorithm.
```
same as other places

File: tianshou/policy/imitation/bcq.py [hunk 10]
```diff
@@ -0,0 +1,326 @@
...
+                .to(self.device).clamp(-0.5, 0.5)
+
+        # decode z with state!
+        return self.max_action * torch.tanh(self.decoder(torch.cat([state, z], 1)))
+
+
+class BCQPolicy(BasePolicy):
+    """Implementation of BCQ algorithm. arXiv:1812.02900.
+
+    :param Perturbation actor: the actor perturbation (s, a -> perturbed a)
```
Comment (Trinkle23897) [inline tianshou/policy/imitation/bcq.py]:
```suggestion
    :param Perturbation actor: the actor perturbation. (s, a -> perturbed a)
```

File: tianshou/policy/imitation/bcq.py [hunk 11]
```diff
@@ -0,0 +1,326 @@
...
+    :param torch.optim.Optimizer critic1_optim: the optimizer for the first
+        critic network.
+    :param torch.nn.Module critic2: the second critic network. (s, a -> Q(s, a))
+    :param torch.optim.Optimizer critic2_optim: the optimizer for the second
+        critic network.
+    :param VAE vae: the VAE network, generating actions similar
+        to those in batch. (s, a -> generated a)
+    :param torch.optim.Optimizer vae_optim: the optimizer for the VAE network.
+    :param Union[str, torch.device] device: which device to create this model on.
+        Default to cpu.
```
Comment (Trinkle23897) [inline tianshou/policy/imitation/bcq.py]:
```suggestion
        Default to "cpu".
```

File: tianshou/policy/imitation/bcq.py [hunk 12]
```diff
@@ -0,0 +1,326 @@
...
+        self,
+        actor: Perturbation,
+        actor_optim: torch.optim.Optimizer,
+        critic1: torch.nn.Module,
+        critic1_optim: torch.optim.Optimizer,
+        critic2: torch.nn.Module,
+        critic2_optim: torch.optim.Optimizer,
+        vae: VAE,
+        vae_optim: torch.optim.Optimizer,
+        device: Optional[Union[str, torch.device]] = "cpu",
```
Comment (Trinkle23897) [inline tianshou/policy/imitation/bcq.py]: why optional?

File: tianshou/policy/imitation/bcq.py [hunk 13]
```diff
@@ -0,0 +1,326 @@
...
+        self,
+        batch: Batch,
+        state: Optional[Union[dict, Batch, np.ndarray]] = None,
+        input: str = "obs",
+        **kwargs: Any,
+    ) -> Batch:
+        """Compute action over the given batch data."""
+        # state: None, input: "obs"
+        # There is "obs" in the Batch
+        # obs: 10 groups. Each group has a state. shape: (10, state_dim)
```
Comment (Trinkle23897) [inline tianshou/policy/imitation/bcq.py]: why 10? Can it be another number?

File: tianshou/policy/imitation/bcq.py [hunk 14]
```diff
@@ -0,0 +1,326 @@
...
+        obs_group = torch.FloatTensor(batch["obs"]).to(self.device)
+
+        act = []
+        with torch.no_grad():
+            for obs in obs_group:
+                # now obs is (state_dim)
+                obs = (obs.reshape(1, -1)).repeat(100, 1)
+                # now obs is (100, state_dim)
```
Comment (Trinkle23897) [inline tianshou/policy/imitation/bcq.py]: why 100?

File: tianshou/policy/imitation/bcq.py [hunk 15]
```diff
@@ -0,0 +1,326 @@
...
+        batch: Batch,
+        state: Optional[Union[dict, Batch, np.ndarray]] = None,
+        input: str = "obs",
+        **kwargs: Any,
+    ) -> Batch:
+        """Compute action over the given batch data."""
+        # state: None, input: "obs"
+        # There is "obs" in the Batch
+        # obs: 10 groups. Each group has a state. shape: (10, state_dim)
+        obs_group = torch.FloatTensor(batch["obs"]).to(self.device)
```
Comment (Trinkle23897) [inline tianshou/policy/imitation/bcq.py]: Can we do this?
```suggestion
        obs_group = to_torch(batch["obs"], device=self.device)
```

File: tianshou/policy/imitation/bcq.py [hunk 16]
```diff
@@ -0,0 +1,326 @@
...
+        # (batch_size, state_dim)
+        obs = torch.FloatTensor(batch["obs"]).to(device=self.device)
+        # (batch_size, action_dim)
+        act = torch.FloatTensor(batch["act"]).to(device=self.device)
+        # (batch_size)
+        rew = torch.FloatTensor(batch["rew"]).to(device=self.device)
+        # (batch_size)
+        done = torch.IntTensor(batch["done"]).to(device=self.device)
+        # (batch_size, state_dim)
+        obs_next = torch.FloatTensor(batch["obs_next"]).to(device=self.device)
```
Comment (Trinkle23897) [inline tianshou/policy/imitation/bcq.py]: same, for conversion you can refer to other policy's approaches instead of manually handling this (e.g., what if obs is not a numpy array?)

File: tianshou/policy/imitation/bcq.py [hunk 17]
```diff
@@ -0,0 +1,326 @@
...
+        KL_loss = (-torch.log(std) + (std.pow(2) + mean.pow(2) - 1) / 2).mean()
+        vae_loss = recon_loss + KL_loss / 2
+
+        self.vae_optim.zero_grad()
+        vae_loss.backward()
+        self.vae_optim.step()
+
+        # critic training:
+        with torch.no_grad():
+            obs_next = obs_next.repeat_interleave(10, dim=0)  # repeat 10 times
```
Comment (Trinkle23897) [inline tianshou/policy/imitation/bcq.py]: why 10?

File: test/continuous/test_bcq.py (hunk 5 above)
Comment (Trinkle23897) [reply test/continuous/test_bcq.py]: for example https://github.com/RKrahl/pytest-dependency

File: tianshou/policy/imitation/bcq.py (hunk 13 above)
Comment (thkkk) [reply tianshou/policy/imitation/bcq.py]: I forgot to modify it when writing this code. I think it is the training_num or test_num. I'll modify it.

File: examples/offline/offline_bcq.py (hunk 2 above)
Comment (thkkk) [reply examples/offline/offline_bcq.py]: Yes. I think VAE may be useful for other algorithms.

File: examples/offline/offline_bcq.py (hunk 3 above)
Comment (thkkk) [reply examples/offline/offline_bcq.py]: In supplementary material of [BCQ paper](https://arxiv.org/pdf/1812.02900.pdf)(Page 22), the dimensionality of z is set to twice the dimensionality of the action space. I'll add an arg in `offline_bcq.py`.

File: tianshou/policy/imitation/bcq.py [hunk 18]
```diff
@@ -0,0 +1,213 @@
...
+
+            # decode(obs) generates action and actor perturbs it
+            action = self.actor(obs, self.vae.decode(obs))
+            # now action is (forward_sampled_times, action_dim)
+            q1 = self.critic1(obs, action)
+            # q1 is (forward_sampled_times, 1)
+            ind = q1.argmax(0)
+            act.append(action[ind].cpu().data.numpy().flatten())
+        act = np.array(act)
+        return Batch(act=act)
```
Comment (Trinkle23897) [inline tianshou/policy/imitation/bcq.py:122]: Can we batch these operations instead of using for-loop?

**Review by Trinkle23897 — DISMISSED** [review]

File: tianshou/policy/imitation/bcq.py (hunk 18 above)
Comment (Trinkle23897) [reply tianshou/policy/imitation/bcq.py:122]: this is because in MLP it uses `flatten(1)` instead of `reshape(-1, dim)`

Comment (thkkk) [reply tianshou/policy/imitation/bcq.py:122]: like this:
```python
obs: torch.Tensor = to_torch(
    batch.obs, device=self.device
)
obs = obs.reshape(-1, 1, obs.shape[1]).repeat(1, self.forward_sampled_times, 1)
act = self.actor(obs, self.vae.decode(obs))
q1 = self.critic1(obs, act)
ind = q1.argmax(1)
act = act[range(ind.shape[0]), ind.reshape(-1)].cpu().data.numpy()
```

Comment (thkkk) [reply tianshou/policy/imitation/bcq.py:122]: But this requires modification of the MLP and Critic. I think they should fit more dimensional input.

Comment (Trinkle23897) [reply tianshou/policy/imitation/bcq.py:122]: will do in the following PR

**Review by Trinkle23897 — APPROVED** [review]
